<!doctype html>
<html lang="en">
<link rel="icon" type="image/png" href="/assets/img/favicon.png" >
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="search-domain" value="http://localhost:4000/">
    <link href="https://fonts.googleapis.com/css?family=Raleway" rel="stylesheet">
    <link rel="stylesheet" href="/assets/css/syntax.css">
    <link rel="stylesheet" type="text/css" href="/assets/css/bootstrap.css" />
    <link rel="stylesheet" type="text/css" href="/assets/css/bootstrap-theme.css" />
    <link rel="stylesheet" type="text/css" href="/assets/css/main.css" />
    <title>Bastien Boussau's github website: Two projects using RevBayes</title>
  </head>
  <body>
    <div class="container">
      <nav class="navbar navbar-default navbar-fixed-top">
  <div class="container-fluid">
    <div class="navbar-header">
      <a href="/" class="pull-left">
        
        <img class="navbar-logo" src="/assets/img/aquabayes-desaturated.png" alt="RevBayes Home" />
        
      </a>

      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar" align="right"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>

    </div>
    <div class="collapse navbar-collapse" id="bs-example-navbar-collapse-1">
      <ul class="nav navbar-nav">
      <li><a href="/tutorials/">Tutorials</a></li>
      <li><a href="/tutorials/relative_time_constraints/">Dating with relative constraints</a></li>
      </ul>
      <!-- <form class="navbar-form navbar-right" role="search" id="search" onsubmit="google_search(); return false;">
        <div class="form-group">
          <input type="text" id="google-search" placeholder="Search..." aria-label="Google site search">
        </div>
      </form> -->
    </div>
  </div>
</nav>

      <div class="titlebar">
	<h1 class="maintitle">Two projects using RevBayes</h1>
	<h3 class="subtitle"></h3>
	<h4 class="authors">Nicolas Lartillot, Bastien Boussau</h4>
  <h5>Last modified on </h5>
</div>


<div class="sidebar no-print">
<blockquote class="overview" id="overview">
  <h2>Overview</h2>
  
  <div class="row">
    <div class="col-md-9">
        <strong>Prerequisites</strong>
        
          <ul id="prerequisites">
          
            <li><a href="/tutorials/M1_Tutorials/EvolMolTP1/">Evol Mol TP1a - A primer on Bayesian phylogenetic models</a></li>
          
            <li><a href="/tutorials/M1_Tutorials/EvolMolTP2/">Evol Mol TP1b - Nucleotide substitution models</a></li>
          
            <li><a href="/tutorials/M1_Tutorials/EvolMolTP3/">Evol Mol TP1c - Codon models</a></li>
          
          </ul>
        
    </div>
  </div>
  
</blockquote>





<blockquote class="tutorial_files" id="tutorial_files">
    <h2>Data files and scripts</h2>
    
        
        <strong>Data Files</strong>
        <ul id="data_files">
        
        
        
          <li><a href="/tutorials/M1_Tutorials/EvolMolProjects/data/gam30status.nex">gam30status.nex</a></li>
        
          <li><a href="/tutorials/M1_Tutorials/EvolMolProjects/data/gamprot30.tree">gamprot30.tree</a></li>
        
          <li><a href="/tutorials/M1_Tutorials/EvolMolProjects/data/primlht.nex">primlht.nex</a></li>
        
          <li><a href="/tutorials/M1_Tutorials/EvolMolProjects/data/simians.tree">simians.tree</a></li>
        
          <li><a href="/tutorials/M1_Tutorials/EvolMolProjects/data/simians4fold.nex">simians4fold.nex</a></li>
        
          <li><a href="/tutorials/M1_Tutorials/EvolMolProjects/data/tax30COG49_rpsG.nex">tax30COG49_rpsG.nex</a></li>
        
          <li><a href="/tutorials/M1_Tutorials/EvolMolProjects/data/tax30COG51_rpsJ.nex">tax30COG51_rpsJ.nex</a></li>
        
          <li><a href="/tutorials/M1_Tutorials/EvolMolProjects/data/tax30COG52_rpsB.nex">tax30COG52_rpsB.nex</a></li>
        
        </ul>
    
        
        <strong>Scripts</strong>
        <ul id="scripts">
        
        
        
          <li><a href="/tutorials/M1_Tutorials/EvolMolProjects/scripts/gam30omega.rev">gam30omega.rev</a></li>
        
          <li><a href="/tutorials/M1_Tutorials/EvolMolProjects/scripts/prim_strict_clock.rev">prim_strict_clock.rev</a></li>
        
        </ul>
    
</blockquote>


</div>
<h2 class="section" id="a-relaxed-molecular-clock-informed-by-life-history-and-germ-line-development-in-primatespnas2016amsterpdf">A relaxed molecular clock informed by <a href="PNAS2016Amster.pdf">life-history and germ-line development in primates</a></h2>
<hr class="section" />

<p>In primates, the mutation rate per generation is mostly determined by the number of replications in the germline. The developmental process of the germline is relatively well characterized, and a model of its modulation, as a function of life-history (age of puberty, generation time, etc) has been proposed (Amster and Sella, 2016). Here, the aim is to use this model (and the empirically measured values reported in the article of Amster and Sella) to calibrate the molecular clock (and its variation) in simian primates, using information about sexual maturity and generation time in extant species.</p>

<p>As a starting point, a script is provided, which implements a strict clock analysis using a multiple sequence alignment (only the 3d coding positions) for simians. This dating analysis is calibrated based on the mutation rate estimated in Humans. In addition, the program also models the evolution of the generation time along the tree, using a Gaussian model. You can run, and read, this script to understand its structure. Then, the idea is to use information about generation times along the tree to modulate the molecular clock along the branches, according to a (simplified version of) the mechanistic model suggested by Amster and Sella.</p>

<h2 class="section" id="analysis-of-gene-alignments-from-gamma-proteobacteria">Analysis of gene alignments from Gamma Proteobacteria</h2>
<hr class="section" />

<p>Here the project is to analyze alignments from several genes from gamma-Proteobacteria, some of which are free-living, and others are endosymbionts. We want to test whether selection is less efficient in endosymbiotic species, owing to their smaller effective population sizes.</p>

<p>Three single gene alignments for 30 bacteria are given, together with a pre-estimated tree and a file that contains the information about the status (free-living or endosymbiotic) for each of the 30 species. Finally, a script is given, gam30omega.rev, which implements a simple codon model that assumes the same dN/dS over all branches of the tree. Starting from this script, the idea is to implement a model that would assume a different dN/dS for the two types of bacteria.</p>


<ol class="bibliography"></ol>

<script type="text/javascript">
var _ol = document.querySelectorAll('ol');
for (var i = 0, elem_ol; elem_ol = _ol[i]; i++) {
	if ( elem_ol.classList == "bibliography" ) {
		var _li = elem_ol.getElementsByTagName("li");
		//for (var j = 0, elem_li; elem_li = _li[j]; j++)
		//{
		//	elem_li.innerHTML = elem_li.innerHTML.replace(/(https?:\/\/)([^\s<]+)/,"<a href=\"$1$2\">$2");
		//}
		if(_li.length > 0)
			elem_ol.outerHTML = "<h2 class='references'>References</h2><hr class='references'>"+elem_ol.outerHTML
	}
}
</script>

      <br>
<footer>
  <div class="container">
  <div class="row">
    <div class="col-sm-12" align="center">
      <a href="https://github.com/revbayes">GitHub</a>  | <a href="https://groups.google.com/forum/#!forum/revbayes-users">Users Forum</a>
    </div>
  </div>
  <br>
  </div>
</footer>

    </div>
    <script src="/assets/js/vendor/jquery.min.js"></script>
<script src="/assets/js/vendor/FileSaver.min.js"></script>
<script src="/assets/js/vendor/jszip.min.js"></script>
<script src="/assets/js/vendor/bootstrap.min.js"></script>

<script type="text/javascript">
// Add default language
$(":not(code).highlighter-rouge").each(function() {
  
  if( this.classList == "highlighter-rouge") {
    this.classList = "Rev highlighter-rouge";
  }
  
});
// $("code.highlighter-rouge").each(function() {
//   
//   if( this.classList == "highlighter-rouge") {
//       this.classList = "Rev highlighter-rouge";
//   }
//   
// });
</script>
<script type="text/javascript"
    src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
    MathJax.Hub.Config({
    tex2jax: {
      inlineMath: [['$','$'], ['\\(','\\)']],
      processEscapes: true
    }
    });
    MathJax.Hub.Queue(function () {
      $(".aside").each(function() {
          $("div .MathJax", this).hide();
      });
    });
</script>
<script src="/assets/js/base.js"></script>

  </body>
</html>
