<!doctype html>
<html lang="en">
<link rel="icon" type="image/png" href="/assets/img/favicon.png" >
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="search-domain" value="https://boussau.github.io/">
    <link href="https://fonts.googleapis.com/css?family=Raleway" rel="stylesheet">
    <link rel="stylesheet" href="/assets/css/syntax.css">
    <link rel="stylesheet" type="text/css" href="/assets/css/bootstrap.css" />
    <link rel="stylesheet" type="text/css" href="/assets/css/bootstrap-theme.css" />
    <link rel="stylesheet" type="text/css" href="/assets/css/main.css" />
    <title>Bastien Boussau's github website: Projects using RevBayes</title>
  </head>
  <body>
    <div class="container">
      <nav class="navbar navbar-default navbar-fixed-top">
  <div class="container-fluid">
    <div class="navbar-header">
      <a href="/" class="pull-left">
        
        <img class="navbar-logo" src="/assets/img/aquabayes-desaturated.png" alt="RevBayes Home" />
        
      </a>

      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar" align="right"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>

    </div>
    <div class="collapse navbar-collapse" id="bs-example-navbar-collapse-1">
      <ul class="nav navbar-nav">
      <li><a href="/tutorials/">Tutorials</a></li>
      <li><a href="/tutorials/relative_time_constraints/">Dating with relative constraints</a></li>
      </ul>
      <!-- <form class="navbar-form navbar-right" role="search" id="search" onsubmit="google_search(); return false;">
        <div class="form-group">
          <input type="text" id="google-search" placeholder="Search..." aria-label="Google site search">
        </div>
      </form> -->
    </div>
  </div>
</nav>

      <div class="titlebar">
	<h1 class="maintitle">Projects using RevBayes</h1>
	<h3 class="subtitle"></h3>
	<h4 class="authors">Nicolas Lartillot, Bastien Boussau</h4>
  <h5>Last modified on September 22, 2022</h5>
</div>


<div class="sidebar no-print">
<blockquote class="overview" id="overview">
  <h2>Overview</h2>
  
  <div class="row">
    <div class="col-md-9">
        <strong>Prerequisites</strong>
        
          <ul id="prerequisites">
          
            <li><a href="/tutorials/M2_Tutorials/TP2_JukesCantorHumanChimp/">TP1 Implementing Bayesian inference with RevBayes</a></li>
          
            <li><a href="/tutorials/M2_Tutorials/TP3_JC_PhylogenyReconstruction/">TP2 Phylogenetic reconstruction on nucleotide data</a></li>
          
            <li><a href="/tutorials/M2_Tutorials/TP4_SubstitutionModels/">TP3 Nucleotide substitution models</a></li>
          
            <li><a href="/tutorials/M2_Tutorials/TP6_CodonModels/">TP4 Codon models</a></li>
          
            <li><a href="/tutorials/M2_Tutorials/TP7_RelaxedClock/">TP5 Molecular dating and relaxed clock models</a></li>
          
          </ul>
        
    </div>
  </div>
  
</blockquote>





<blockquote class="tutorial_files" id="tutorial_files">
    <h2>Data files and scripts</h2>
    
        
        <strong>Data Files</strong>
        <ul id="data_files">
        
        
        
          <li><a href="/tutorials/Projects/data/Daphnia.tgz">Daphnia.tgz</a></li>
        
          <li><a href="/tutorials/Projects/data/Endosymbionts.tgz">Endosymbionts.tgz</a></li>
        
          <li><a href="/tutorials/Projects/data/Groussin2011.zip">Groussin2011.zip</a></li>
        
          <li><a href="/tutorials/Projects/data/Pisani2015.zip">Pisani2015.zip</a></li>
        
          <li><a href="/tutorials/Projects/data/SEMG2.tgz">SEMG2.tgz</a></li>
        
          <li><a href="/tutorials/Projects/data/Saclier2018.zip">Saclier2018.zip</a></li>
        
          <li><a href="/tutorials/Projects/data/Zhang2014.zip">Zhang2014.zip</a></li>
        
          <li><a href="/tutorials/Projects/data/birds.tgz">birds.tgz</a></li>
        
          <li><a href="/tutorials/Projects/data/fishes.tgz">fishes.tgz</a></li>
        
          <li><a href="/tutorials/Projects/data/mammals.tgz">mammals.tgz</a></li>
        
          <li><a href="/tutorials/Projects/data/simians.tgz">simians.tgz</a></li>
        
          <li><a href="/tutorials/Projects/data/thermus.tgz">thermus.tgz</a></li>
        
        </ul>
    
</blockquote>


</div>
<h2 class="section" id="coevolution-between-substitution-rates-and-life-history-traitsevolution2012lartillot-2pdf">Coevolution between <a href="Evolution2012Lartillot-2.pdf">substitution rates and life-history traits</a></h2>
<hr class="section" />

<p>In mammals, there is substantial variation in substitution rate between lineages. This variation may be correlated with life-history traits: small mammals, with short generation times, tend to evolve faster. Here, the aim is measure the correlation between substitution rate and body mass or other life-history traits in mammals, by building a model of the joint evolutionary process of the substitution rate and life-history traits across branches, using a variant of the Brownian model used for modelling the auto-correlated molecular clock (<a href="Evolution2012Lartillot-2.pdf">Lartillot and Delsuc, 2012</a>).</p>

<h2 class="section" id="a-relaxed-molecular-clock-informed-by-life-history-and-germ-line-development-in-primatespnas2016amsterpdf">A relaxed molecular clock informed by <a href="PNAS2016Amster.pdf">life-history and germ-line development in primates</a></h2>
<hr class="section" />

<p>In primates, the mutation rate per generation is mostly determined by the number of replications in the germline. The developmental process of the germline is relatively well characterized, and a model of its modulation, as a function of life-history (age of puberty, generation time, etc) has been proposed (Amster and Sella, 2016). Here, the aim is to use this model (and the empirically measured values reported in the article of Amster and Sella) to calibrate the molecular clock (and its variation) in simian primates, using information about sexual maturity and generation time in extant species.</p>

<h2 class="section" id="estimating-the-strength-of-gc-biased-gene-conversionmbe2013lartillot-1pdf">Estimating the strength of <a href="MBE2013Lartillot-1.pdf">GC-biased gene conversion</a></h2>
<hr class="section" />

<p>The substitution patterns at neutral nucleotide positions in mammalian genomes is not just determined by the mutational process. Another force, GC biased gene conversion (gBGC) is also impacting substitution rates, by increasing the probability of fixation of G and C, compared to A and T. Using population genetics theory, it is possible to derive the nucleotide substitution rate matrix, as a function of the mutation rates and the intensity of gBGC (Lartillot, 2013). The aim is to develop this model and apply it to primate nucleotide data, so as to estimate the intensity of gBGC (and possibly, its variation between branches) using a nucleotide sequence alignment.</p>

<h2 class="section" id="nucleotide-compositional-variation-across-species-and-its-impact-on-phylogenetic-reconstructionsystbiol2004foster-1pdf">Nucleotide <a href="SystBiol2004Foster-1.pdf">compositional variation across species and its impact on phylogenetic reconstruction</a></h2>
<hr class="section" />

<p>The models that we have considered in the tutorials are homogeneous across branches. As a result, they predict the same nucleotide composition in all species. In practice, this is not the case. In some extreme situations, not accounting for compositional variation can result in phylogenetic reconstruction artifacts (typically, unrelated species with similar compositional biases artifactually cluster together). A simple but particularly striking example is shown in <a href="SystBiol2004Foster-1.pdf">Foster, 2004</a>. In this article, a model with branch specific nucleotide composition is  introduced to improve phylogenetic reconstruction in such cases. Since then, <a href="Heaps_2014.pdf">Heaps et al.</a> have proposed improved models. Here, the aim is to design versions of these models with branch-specific equilibrium frequencies, and to see if they improve phylogenetic reconstruction.</p>

<h2 class="section" id="multi-gene-phylogenetic-reconstruction-of-the-phylogeny-of-birdsjarvis_2014pdf"><a href="Jarvis_2014.pdf">Multi-gene phylogenetic reconstruction of the phylogeny of birds</a></h2>
<hr class="section" />

<p>Designing a model for doing multi-gene phylogenetic reconstruction. Genes may share the same species phylogeny but may differ in their rate of evolution and in their GC content. Can be used to reconstruct the phylogeny or to estimate the variance in substitution rates and in GC content across genes. The data is available from <a href="Jarvis_2014.pdf">the article</a></p>

<h2 class="section" id="correlation-between-gc-composition-of-ribosomal-rna-and-growth-temperature-in-archaeagroussin_2011pdf"><a href="Groussin_2011.pdf">Correlation between GC composition of ribosomal RNA and growth temperature in Archaea</a>.</h2>
<hr class="section" />

<p>The idea is to model the correlated evolution of rRNA GC content and growth temperature across Archaea, and use this model to estimate the correlation and to infer ancestral temperatures along the phylogeny.</p>

<h2 class="section" id="are-patterns-of-absencepresence-of-genes-across-genomes-informative-about-the-phylogeny">Are patterns of absence/presence of genes across genomes informative about the phylogeny?</h2>
<hr class="section" />

<p>The idea is to model the process of gain and loss of genes across a phylogeny, and to apply this model to data of absence/presence of genes across metazoans. See article of <a href="Pisani_2015.pdf">Pisani et al., (2015)</a> and <a href="Ryan_2013.pdf">Ryan et al (2013)</a> for two analyses giving different results on the same dataset.</p>

<h2 class="section" id="convergent-evolution-toward-subterranean-lifestyle-in-isopods-and-its-consequences-on-the-rate-of-genome-evolutionsaclier_2018pdf"><a href="Saclier_2018.pdf">Convergent evolution toward subterranean lifestyle in isopods, and its consequences on the rate of genome evolution</a>.</h2>
<hr class="section" />

<p>Saclier et al have analysed the evolutionary patterns in a group of isopod species, in which there has been a large number of independent transitions from surface to underground lifestyle – giving an opportunity for modeling and investigating the impact of these transitions on genomic sequence evolution.</p>

<h2 class="section" id="exploring-the-feasibility-of-importance-samplinghttpsenwikipediaorgwikiimportance_sampling-for-phylogenetic-inference">Exploring the feasibility of <a href="https://en.wikipedia.org/wiki/Importance_sampling">importance sampling</a> for phylogenetic inference</h2>
<hr class="section" />

<p>Bayesian phylogenetic inference classically relies on MCMC, a computationally intensive algorithm. This is a problem as data sets have been increasing in size, resulting in increased computational and environmental footprints. In this project, you will investigate an alternative algorithm, <a href="https://en.wikipedia.org/wiki/Importance_sampling">importance sampling</a>, as follows. Firstly, you will perform MCMC inference on a single gene alignment, and on a big alignment. Secondly, you will subsample a limited number of samples from the posterior distribution obtained on the small alignment. Thirdly, you will evaluate the posterior probability of these samples according to the big alignment. Fourthly, you will reweigh these samples according to the ratio of their posterior probabilities, on the large data set <em>vs</em> the small one. As a result, on the big dataset, you will then have a posterior distribution obtained through importance sampling, and one obtained using MCMC. On the big dataset, does the faster importance sampling approach produce estimates that are similar to the MCMC approach?</p>

<h2 class="section" id="analysis-of-the-semg2-gene-in-primates">Analysis of the SEMG2 gene in Primates</h2>
<hr class="section" />

<p>Here the project is to analyze an alignment of the SEMG2 gene from primate species that differ in their mating systems.</p>

<h2 class="section" id="analysis-of-mitochondrial-protein-evolution-in-daphnia">Analysis of mitochondrial protein evolution in Daphnia</h2>
<hr class="section" />

<p>Here the project is to analyze a concatenated alignment of 15 DNA sequences coding for proteins from 29 strains of <em>Daphnia pulex</em>, some of which reproduce sexually (named S1 to S14), and others, asexually (named A1 to A14). Sexual reproduction is assumed to be the ancestral condition.</p>

<ol class="bibliography"></ol>

<script type="text/javascript">
var _ol = document.querySelectorAll('ol');
for (var i = 0, elem_ol; elem_ol = _ol[i]; i++) {
	if ( elem_ol.classList == "bibliography" ) {
		var _li = elem_ol.getElementsByTagName("li");
		//for (var j = 0, elem_li; elem_li = _li[j]; j++)
		//{
		//	elem_li.innerHTML = elem_li.innerHTML.replace(/(https?:\/\/)([^\s<]+)/,"<a href=\"$1$2\">$2");
		//}
		if(_li.length > 0)
			elem_ol.outerHTML = "<h2 class='references'>References</h2><hr class='references'>"+elem_ol.outerHTML
	}
}
</script>

      <br>
<footer>
  <div class="container">
  <div class="row">
    <div class="col-sm-12" align="center">
      <a href="https://github.com/revbayes">GitHub</a>  | <a href="https://groups.google.com/forum/#!forum/revbayes-users">Users Forum</a>
    </div>
  </div>
  <br>
  </div>
</footer>

    </div>
    <script src="/assets/js/vendor/jquery.min.js"></script>
<script src="/assets/js/vendor/FileSaver.min.js"></script>
<script src="/assets/js/vendor/jszip.min.js"></script>
<script src="/assets/js/vendor/bootstrap.min.js"></script>

<script type="text/javascript">
// Add default language
$(":not(code).highlighter-rouge").each(function() {
  
  if( this.classList == "highlighter-rouge") {
    this.classList = "Rev highlighter-rouge";
  }
  
});
// $("code.highlighter-rouge").each(function() {
//   
//   if( this.classList == "highlighter-rouge") {
//       this.classList = "Rev highlighter-rouge";
//   }
//   
// });
</script>
<script type="text/javascript"
    src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
    MathJax.Hub.Config({
    tex2jax: {
      inlineMath: [['$','$'], ['\\(','\\)']],
      processEscapes: true
    }
    });
    MathJax.Hub.Queue(function () {
      $(".aside").each(function() {
          $("div .MathJax", this).hide();
      });
    });
</script>
<script src="/assets/js/base.js"></script>

  </body>
</html>
