<!doctype html>
<html lang="en">
<link rel="icon" type="image/png" href="/assets/img/favicon.png" >
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="search-domain" value="https://boussau.github.io/">
    <link href="https://fonts.googleapis.com/css?family=Raleway" rel="stylesheet">
    <link rel="stylesheet" href="/assets/css/syntax.css">
    <link rel="stylesheet" type="text/css" href="/assets/css/bootstrap.css" />
    <link rel="stylesheet" type="text/css" href="/assets/css/bootstrap-theme.css" />
    <link rel="stylesheet" type="text/css" href="/assets/css/main.css" />
    <title>Bastien Boussau's github website: Evol Mol TP2 - Nucleotide substitution models</title>
  </head>
  <body>
    <div class="container">
      <nav class="navbar navbar-default navbar-fixed-top">
  <div class="container-fluid">
    <div class="navbar-header">
      <a href="/" class="pull-left">
        
        <img class="navbar-logo" src="/assets/img/aquabayes-desaturated.png" alt="RevBayes Home" />
        
      </a>

      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar" align="right"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>

    </div>
    <div class="collapse navbar-collapse" id="bs-example-navbar-collapse-1">
      <ul class="nav navbar-nav">
      <li><a href="/tutorials/">Tutorials</a></li>
      <li><a href="/tutorials/relative_time_constraints/">Dating with relative constraints</a></li>
      </ul>
      <!-- <form class="navbar-form navbar-right" role="search" id="search" onsubmit="google_search(); return false;">
        <div class="form-group">
          <input type="text" id="google-search" placeholder="Search..." aria-label="Google site search">
        </div>
      </form> -->
    </div>
  </div>
</nav>

      <div class="titlebar">
	<h1 class="maintitle">Evol Mol TP2 - Nucleotide substitution models</h1>
	<h3 class="subtitle">Estimating substitution rates between nucleotides</h3>
	<h4 class="authors">Nicolas Lartillot</h4>
  <h5>Last modified on November 15, 2021</h5>
</div>


<div class="sidebar no-print">
<blockquote class="overview" id="overview">
  <h2>Overview</h2>
  
  <div class="row">
    <div class="col-md-9">
        <strong>Prerequisites</strong>
        
          <ul id="prerequisites">
          
            <li><a href="/tutorials/L3_Tutorials/EvolMolTP1/">Evol Mol TP1 - Phylogenetic reconstruction on nucleotide data</a></li>
          
          </ul>
        
    </div>
  </div>
  
</blockquote>





<blockquote class="tutorial_files" id="tutorial_files">
    <h2>Data files and scripts</h2>
    
        
        <strong>Data Files</strong>
        <ul id="data_files">
        
        
        
          <li><a href="/tutorials/L3_Tutorials/EvolMolTP2/data/placBRCA1short.nex">placBRCA1short.nex</a></li>
        
          <li><a href="/tutorials/L3_Tutorials/EvolMolTP2/data/placBRCA1third.nex">placBRCA1third.nex</a></li>
        
        </ul>
    
        
        <strong>Scripts</strong>
        <ul id="scripts">
        
        
        
          <li><a href="/tutorials/L3_Tutorials/EvolMolTP2/scripts/phyloJC.rev">phyloJC.rev</a></li>
        
        </ul>
    
</blockquote>


</div>
<p>The last tutorial was introducing how to conduct a Bayesian phylogenetic analysis under the Jukes-Cantor model. In this tutorial, we will follow up on this, by exploring other models of nucleotide substitution.</p>

<p>A model of nucleotide substitution specifies the rates of substitution between all pairs of nucleotides. It is entirely defined by a 4x4 matrix called instantaneous rate matrix, or more simply, rate matrix ($Q$).</p>

<p>We have seen in the last tutorial that the instantaneous rate matrix for the JC model is given by:</p>

\[Q_{JC} = \begin{pmatrix}
{*} &amp; \frac{1}{3} &amp; \frac{1}{3} &amp; \frac{1}{3} \\
\frac{1}{3} &amp; {*} &amp; \frac{1}{3} &amp; \frac{1}{3} \\
\frac{1}{3} &amp; \frac{1}{3} &amp; {*} &amp; \frac{1}{3} \\
\frac{1}{3} &amp; \frac{1}{3} &amp; \frac{1}{3} &amp; {*}
\end{pmatrix} \mbox{  ,}\]

<p>This model, however, is simplistic. In practice, different types of substitutions (or different types of point mutations) might not occur at the same rate. Thus, we might need to use more sophisticated models.</p>

<p>Here, we will explore 3 alternative models, of increasing complexity. For each substitution model, we will write a script and run it. This will give us a joint posterior distribution on trees, but also on model parameters. Once the analysis has been conducted, we can then interpret the posterior distribution over the parameters of the model. This will give us information about the process of molecular evolution.</p>

<p>The models explored in this tutorial are summarized in this table.</p>

<figure id="tab_subst_models" class="table"><figcaption class="table">Specific functions for substitution models available in RevBayes.</figcaption>

<table>
  <thead>
    <tr>
      <th style="text-align: center"><strong>Model</strong></th>
      <th style="text-align: center"><strong>Reference</strong></th>
      <th style="text-align: center"><strong>Function</strong></th>
      <th style="text-align: center"><strong>Parameters</strong></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="text-align: center">Jukes-Cantor</td>
      <td style="text-align: center"><a class="citation" href="#Jukes1969">(Jukes and Cantor 1969)</a></td>
      <td style="text-align: center">fnJC</td>
      <td style="text-align: center">-</td>
    </tr>
    <tr>
      <td style="text-align: center">T92</td>
      <td style="text-align: center"><a class="citation" href="#Tamura1992">(Tamura 1992)</a></td>
      <td style="text-align: center">fnT92</td>
      <td style="text-align: center">$\pi_{GC}$, $\kappa$</td>
    </tr>
    <tr>
      <td style="text-align: center">HKY</td>
      <td style="text-align: center"><a class="citation" href="#Hasegawa1985">(Hasegawa et al. 1985)</a></td>
      <td style="text-align: center">fnHKY</td>
      <td style="text-align: center">$\pi$, $\kappa$</td>
    </tr>
    <tr>
      <td style="text-align: center">GTR</td>
      <td style="text-align: center"><a class="citation" href="#Tavare1986">(Tavar√© 1986)</a></td>
      <td style="text-align: center">fnGTR</td>
      <td style="text-align: center">$\pi$, $\epsilon$</td>
    </tr>
  </tbody>
</table>
</figure>
<p>They are now presented in more details.</p>

<h2 class="section" id="the-tamura-92-model-t92">The Tamura 92 model (T92)</h2>
<hr class="section" />

<h3 class="subsection" id="description-of-the-model">Description of the model</h3>
<hr class="subsection" />

<p>This model has two parameters, $\kappa$ and $\gamma$. The rate matrix is given by:</p>

\[Q_{T92} = \begin{pmatrix}
{*} &amp; \frac{\gamma}{2} &amp; \kappa \frac{\gamma}{2} &amp; \frac{1-\gamma}{2} \\
\frac{1-\gamma}{2} &amp; {*} &amp; \frac{\gamma}{2} &amp; \kappa \frac{1-\gamma}{2} \\
\kappa \frac{1-\gamma}{2} &amp; \frac{\gamma}{2} &amp; {*} &amp; \frac{1-\gamma}{2} \\
\frac{1-\gamma}{2} &amp; \kappa \frac{\gamma}{2} &amp; \frac{\gamma}{2} &amp; {*}
\end{pmatrix} \mbox{  ,}\]

<p>Two things are important to note:</p>
<ul>
  <li>if $\kappa &gt; 1$, then this says that transitions between purines (A and G) or between pyrimidines (C and T) are more frequent than transversions (from a purine to a pyrimidine, or vice versa). If $kappa &lt; 1$, on the other hand, then transitions are less frequent than transversions.</li>
  <li>if $\gamma &lt; 0.5$, then substitutions from GC to AT occur at a higher rate, compared to substitutions from AT to GC, and conversely if $\gamma &gt; 0.5$.</li>
</ul>

<p>For which values of $\kappa$ and $\gamma$ does the T92 model reduce to JC ?</p>

<h3 class="subsection" id="implementation">Implementation</h3>
<hr class="subsection" />

<p>A script (<code class="highlighter-rouge">scripts/phyloJC.rev</code>), implementing phylogenetic inference with the JC model, is given with the tutorial. It is slightly different from the one that you used for the last tutorial, but is essentially equivalent. It was just re-written so as to better separate the model declaration from the moves and the monitors. In addition, the move schedule was slightly modified, so as to allow for faster analysis on the specific dataset of interest for today. This dataset is an alignment of the BRCA1 gene in placental mammals (<code class="highlighter-rouge">data/placBRCA1.nex</code>). To further accelerate the analyses, we will in fact use a short version of this dataset (in the file called <code class="highlighter-rouge">placBRCA1short.nex</code>), which contains only the 600 first nucleotide positions of this gene. Finally, a third dataset is provided (file called <code class="highlighter-rouge">placBRCA1third.nex</code>), which contains only the third coding positions of the original alignment. In the following, we will explore both <code class="highlighter-rouge">placBRCA1short.nex</code> and <code class="highlighter-rouge">placBRCA1third.nex</code>.</p>

<p>Copy the JC script into a new script, which you may call <code class="highlighter-rouge">phyloT92.rev</code>), and open it. You will modify this script so as to implement a T92 model. Much of the script should stay the same, but you will have to introduce several modifications in order to take into account the specific aspects of this new and more complex model.</p>

<p>First, under JC, the rate matrix and the sequence evolutionary process were specified as follows:</p>
<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Q &lt;- fnJC(4)
</code></pre></div></div>
<p>In the case of JC, the rate matrix is a constant, hence the use of <code class="highlighter-rouge">&lt;-</code>.</p>

<p>Then, we create a stochastic variable, which represents the multiple sequence alignment, such as produced by a substitution process running along the tree psi, and with substitution rates given by $Q$:</p>
<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>seq ~ dnPhyloCTMC(tree=psi, Q=Q, type="DNA")
</code></pre></div></div>

<p>If we now want to use a T92 model instead of a JC model, we need, first, to defined the two parameters: $\kappa$ and $\gamma$, as stochastic variables having a prior. Then, we can build the $Q$ matrix, and finally, we create a <code class="highlighter-rouge">phyloCTMC</code> (<code class="highlighter-rouge">seq</code>) using this $Q$ matrix. Finally, we need to move these two parameters during the MCMC.</p>

<p>The transition-transversion rate ratio $\kappa$ is a non negative real number. A reasonable prior assumption about $\kappa$ is that it is not much higher than 10. So, we can use an exponential prior of mean 10 (of rate $\lambda = 0.1$).</p>
<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>kappa ~ dnExponential(lambda = 0.1)
</code></pre></div></div>
<p>Concerning $\gamma$, it is a number between 0 and 1. We can invoke a uniform prior distribution. The uniform distribution between 0 and 1 is a particular case of the beta distribution, with with parameters 1.0 and 1.0:</p>
<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>gamma ~ dnBeta(1.0, 1.0)
</code></pre></div></div>
<p>Then, we can create the rate matrix $Q$, which is now a T92 rate matrix:</p>
<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Q := fnT92(kappa = kappa, gc = gamma)
</code></pre></div></div>
<p>Note that the rate matrix is now a deterministic model variable (and not anymore a constant) ‚Äì hence, we use <code class="highlighter-rouge">:=</code>.</p>

<p>Finally, we can proceed with the sequence evolutionary process:</p>
<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>seq ~ dnPhyloCTMC(tree=psi, Q=Q, type="DNA")
</code></pre></div></div>

<p>Concerning the MCMC moves, we need to add moves for $\kappa$ and $\gamma$. For $\kappa$, we can use a scaling (or multiplicative) move, which is the best move to use for positive real numbers. For $\gamma$, we can use a sliding (or additive) move. As in the case of the JC model, we also need to move the tree topology, using NNI and SPR moves, as well as the branch lengths (using scaling moves). Altogether, or move vector can be specified as follows:</p>
<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>moves = VectorMoves()
moves.append(mvNNI(topology, weight=3.0))
moves.append(mvSPR(topology, weight=3.0))
moves.append(mvSlide(gamma, weight=1.0))
moves.append(mvScale(kappa, weight=1.0))
for (i in 1:n_branches) {
   moves.append(mvScale(bl[i], weight=1.0))
}
</code></pre></div></div>

<p>Write the script, run it on the BRCA1 dataset (short version). Load the log file in Tracer. Estimate the burnin and visualize the posterior distribution for $\kappa$. Give a point estimate (median) and a symmetric 95\% credible interval. Do the same thing for $\gamma$.</p>

<p>Based on your point estimated and credible intervals, would you say that T92 provides a better fit to the data, compared to JC?</p>

<p>Do a similar analysis, now with only the third coding positions (file called <code class="highlighter-rouge">placBRCA1third.nex</code>). Run the chain for 10 000 generations. Visualize the posterior distribution and give a point estimate and a credible interval for $\kappa$ and $\gamma$. Compare with the estimates obtained under the complete dataset (all coding positions). How do you interpret the differences?</p>

<p>Print out the map or posterior consensus tree. Does that differ a lot from the consensus tree inferred using the JC model?</p>

<h2 class="section" id="the-hasegawa-kishino-and-yano-model-hky">The Hasegawa, Kishino and Yano model (HKY)</h2>
<hr class="section" />

<p>The HKY model is slightly more complex than T92. It also relies on a transition/transversion ratio ($\kappa$). However, whereas T92 assumes that the equilibrium frequencies of G and C are the same (both equal to $\gamma/2$), and similarly for A and T (with frequencies both equal to $(1-\gamma)/2$, the HKY does not make this assumption. Under HKY, the equilibrium frequencies of the 4 nucleotides can be arbitrary. They are mathematically formalized by a frequency vector of dimension 4, $\pi = (\pi_A, \pi_C, \pi_G, \pi_T)$ such that:
\(\pi_A + \pi_C + \pi_G + \pi_T = 1\)</p>

<p>The rate matrix of HKY is:</p>

\[Q_{HKY} = \begin{pmatrix}
{*} &amp; \pi_{C} &amp; \kappa \pi_{G} &amp; \pi_{T} \\
\pi_{A} &amp; {*}  &amp; \pi_{G} &amp; \kappa \pi_{T} \\
\kappa \pi_{A} &amp; \pi_{C} &amp; {*}  &amp; \pi_{T} \\
\pi_{A} &amp; \kappa \pi_{C} &amp; \pi_{G} &amp; {*}
\end{pmatrix} \mbox{  ,}\]

<p>The T92 model is a particular case of the HKY model. For which values of its parameters does HKY reduce to T92?</p>

<p>The domain of definition of frequency vectors is called the simplex (here, the simplex S4, since there are 4 entries for the vector). A standard distribution on the simplex is the Dirichlet distribution (you can look at the definition of the Dirichlet distribution on Wikipedia). In particular, a uniform distribution over the simplex is a Dirichlet with weights all equal to 1. Thus, we can assume this prior for $\pi$:</p>
<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>pi ~ dnDirichlet([1.0, 1.0, 1.0, 1.0])
</code></pre></div></div>

<p>As for the T92 model, we can assume a broad exponential prior for $\kappa$:</p>
<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>kappa ~ dnexponential(lambda = 0.1)
</code></pre></div></div>
<p>and then define the HKY rate matrix:</p>
<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Q := fnHKY(kappa=kappa, baseFrequencies=pi)
</code></pre></div></div>
<p>and finally, create the substitution process:</p>
<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>seq ~ dnPhyloCTMC(tree=psi, Q=Q, type="DNA")
</code></pre></div></div>

<p>We also need to move $\kappa$ and $\pi$. For $\kappa$, we can use a scaling move, as we did for the T92 model. For $\pi$, we can use a move that preserves the positivity of the entries of the vector, but also the constraint that the 4 entries of the vector should sum to 1. The move that does this is a <code class="highlighter-rouge">DirichletSimplex</code> move. The syntax for this move would be as follows:</p>
<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>moves.append(mvDirichletSimplex(pi, weight=1.0, alpha=10))
</code></pre></div></div>

<p>Write and run the script on the BRCA1 gene (only the third coding positions). Again, visualize the posterior distribution over the model parameters. In particular, visualize simultaneously the posterior distributions over the 4 nucleotide frequencies. Would you say that HKY should be used instead of T92? Is tree inference fundamentally different between the two models?</p>

<h2 class="section" id="the-gtr-model-optional">The GTR model (optional)</h2>
<hr class="section" />

<p>The most general time reversible model can be parameterized as follows:
\(Q = \begin{pmatrix}
{*} &amp; \rho_{AC} \pi_C &amp; \rho_{AG} \pi_G &amp; \rho_{AT} \pi_T \\
\rho_{CA} \pi_A &amp; {*}  &amp; \rho_{CG} \pi_G &amp; \rho_{CT} \pi_T \\
\rho_{GA} \pi_A &amp; \rho_{GC}  \pi_C &amp; {*}  &amp; \rho_{GT} \pi_T \\
\rho_{TA} \pi_A &amp; \rho_{TC} \pi_C &amp; \rho_{TG} \pi_G &amp; {*}
\end{pmatrix} \mbox{  ,}\)
with the constraint that $\rho_{XY} = \rho_{YX}$ for all nucleotide pairs $(X,Y)$. The $\rho$‚Äôs are called the relative exchangeabilities (or exchange rates), and the $\pi$‚Äôs are the equilibrium frequencies of the process. With the constraint of symmetry for $\rho$, we have 6 distinct values (6 unordered pairs of nucleotides). In addition, and without loss of generality, we can constrain the $\rho$ vector to sum to 1. With these constraints, $\rho$ is a 6-dimensional frequency vector (i.e. it is a point on the simplex $S6$).</p>

<p>We can use a uniform prior over $\rho$, which is a particular case of the Dirichlet distribution with equal weights:</p>
<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>rho ~ dnDirichlet([1.0, 1.0, 1.0, 1.0, 1.0, 1.0])
</code></pre></div></div>
<p>and a uniform prior over $\pi$ (as for the HKY model):</p>
<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>pi ~ dnDirichlet([1.0, 1.0, 1.0, 1.0])
</code></pre></div></div>
<p>We can then define the GTR rate matrix:</p>
<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Q := fnGTR(exchangeRates=rho, baseFrequencies=pi)
</code></pre></div></div>
<p>and create the substitution process:</p>
<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>seq ~ dnPhyloCTMC(tree=psi, Q=Q, type="DNA")
</code></pre></div></div>
<p>Finally, we need to move $\rho$ and $\pi$, in both cases using a <code class="highlighter-rouge">DirichletSimplex</code> move (as we did above for pi in the case of the HKY model).</p>

<p>Write the script and run it on the BRCA1 dataset (only the 3 coding positions). Explore the posterior distribution over the parameters and, in particular, determine whether all transitions occur at the same relative rate. Same thing for transitions.</p>

<ol class="bibliography"><li><span id="Hasegawa1985">Hasegawa M., Kishino H., Yano T. 1985. Dating of the Human-Ape Splitting by a molecular Clock of Mitochondrial DNA. Journal of Molecular Evolution. 22:160‚Äì174.</span>

<a href="https://doi.org/10.1007/BF02101694">10.1007/BF02101694</a>

</li>
<li><span id="Jukes1969">Jukes T.H., Cantor C.R. 1969. Evolution of Protein Molecules. Mammalian Protein Metabolism. 3:21‚Äì132.</span>

<a href="https://doi.org/10.1016/B978-1-4832-3211-9.50009-7">10.1016/B978-1-4832-3211-9.50009-7</a>

</li>
<li><span id="Tamura1992">Tamura K. 1992. Estimation of the number of nucleotide substitutions when there are strong transition-transversion and G+ C-content biases. Molecular Biology and Evolution. 9:678‚Äì687.</span>

</li>
<li><span id="Tavare1986">Tavar√© S. 1986. Some Probabilistic and Statistical Problems in the Analysis of DNA Sequences. Some Mathematical Questions in Biology: DNA Sequence Analysis. 17:57‚Äì86.</span>

</li></ol>

<script type="text/javascript">
var _ol = document.querySelectorAll('ol');
for (var i = 0, elem_ol; elem_ol = _ol[i]; i++) {
	if ( elem_ol.classList == "bibliography" ) {
		var _li = elem_ol.getElementsByTagName("li");
		//for (var j = 0, elem_li; elem_li = _li[j]; j++)
		//{
		//	elem_li.innerHTML = elem_li.innerHTML.replace(/(https?:\/\/)([^\s<]+)/,"<a href=\"$1$2\">$2");
		//}
		if(_li.length > 0)
			elem_ol.outerHTML = "<h2 class='references'>References</h2><hr class='references'>"+elem_ol.outerHTML
	}
}
</script>

      <br>
<footer>
  <div class="container">
  <div class="row">
    <div class="col-sm-12" align="center">
      <a href="https://github.com/revbayes">GitHub</a>  | <a href="https://groups.google.com/forum/#!forum/revbayes-users">Users Forum</a>
    </div>
  </div>
  <br>
  </div>
</footer>

    </div>
    <script src="/assets/js/vendor/jquery.min.js"></script>
<script src="/assets/js/vendor/FileSaver.min.js"></script>
<script src="/assets/js/vendor/jszip.min.js"></script>
<script src="/assets/js/vendor/bootstrap.min.js"></script>

<script type="text/javascript">
// Add default language
$(":not(code).highlighter-rouge").each(function() {
  
  if( this.classList == "highlighter-rouge") {
    this.classList = "Rev highlighter-rouge";
  }
  
});
// $("code.highlighter-rouge").each(function() {
//   
//   if( this.classList == "highlighter-rouge") {
//       this.classList = "Rev highlighter-rouge";
//   }
//   
// });
</script>
<script type="text/javascript"
    src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
    MathJax.Hub.Config({
    tex2jax: {
      inlineMath: [['$','$'], ['\\(','\\)']],
      processEscapes: true
    }
    });
    MathJax.Hub.Queue(function () {
      $(".aside").each(function() {
          $("div .MathJax", this).hide();
      });
    });
</script>
<script src="/assets/js/base.js"></script>

  </body>
</html>
